### DISCUSSION TITLE
Cannot find Module imported from node modules

### DISCUSSION THREAD

**Discussion by skoenfaelt** [nitrojs/nitro#2121] [Q&A] [answered]
Hello,
whats the playground
1. Nuxt Apllication in version 3.9
2. Own created Component Lib(from now on: `qs-ui`) with Devextreme and Vuetify based on vue 3
3. in dev mode everything works as it should

```
nuxt app project
└───qs-ui
│   └───devextreme
│       │   devextreme/ui/data_grid
```

----

the QS-UI library has some generalized components from other libraries with some tweaked business logic. 
For example a table component, which has the following import:

(table.vue)
`import { DxDataGrid, DxMasterDetail, DxPager, DxPaging } from 'devextreme-vue/data-grid'`

QS Ui is built with following vite and rollup configs

<details>
  <summary>Open Me to show</summary>
  
  ```javascript
export default defineConfig({
  root: './',
  publicDir: 'public',
  plugins: [
    dts({
      insertTypesEntry: true,
      exclude: ['./src/**/*.{cy,stories}.ts']
    }),
    vue({
      template: { transformAssetUrls }
    }),
    vuetify({
      autoImport: true
    })
  ],
  resolve: {
    alias: {
      '@': fileURLToPath(new URL('./src', import.meta.url))
    }
  },
  optimizeDeps: {
    entries: ['src/components/**/*.{cy,spec}.{ts,tsx}'],
    include: ['vuetify', 'devextreme-vue/data-grid']
  },
  build: {
    cssCodeSplit: true,
    lib: {
      entry: resolve(__dirname, 'src/qs-ui.ts'),
      name: 'qs-ui',
      formats: ['es', 'umd'],
      fileName: 'qs-ui'
    },
    rollupOptions: {
      input: {
        main: resolve(__dirname, 'src/qs-ui.ts')
      },
      external: ['vue', 'vuetify/lib', 'axios', 'devextreme', 'devextreme-vue/data-grid'],
      output: {
        assetFileNames: 'qs-ui.[ext]',
        globals: {
          axios: 'axios',
          vue: 'Vue',
          vuetify: 'Vuetify',
          'vuetify/components': 'VuetifyComponents',
          'vuetify/directives': 'VuetifyDirectives'
        }
      }
    },
    // Reduce bloat from legacy polyfills.
    target: 'esnext',
    sourcemap: true
  }
})
  ```
  
</details>

this package got published in self hosted gitlab registry

----

Nuxt App installs the QS Ui library as a plugin

<details>
  <summary>Open me to show</summary>
  
  ```javascript
  export default defineNuxtPlugin((nuxtApp) => {
  const config = useRuntimeConfig()

  initializeDevextreme()

  nuxtApp.vueApp.use(qsUi, {
    activeTheme: ThemesLightEnum.GREEN_THEME,
    httpOptions: {
      baseURL: config.public.backendUrl
    },
    ssr: true
  })
})
  ```
  
</details>

In the nuxt.config i need to tweak some general build issues via this configs:

```
export default defineNuxtConfig({
  devtools: { enabled: true },
  build: {
    transpile: ['@q-soft/ui']
  },

// ....

  vite: {
    optimizeDeps: {
      include: ['devextreme-vue/data-grid']
    }
  }
})
```

So development build works fine. 

# the issue
As I run yarn build (with nodeLinker: node-modules) the client and server is built correctly. But when building the nitro server I get this error: 

> Error: Cannot find module devextreme/ui/data_grid imported from file:///C:/Users/`path/to/project`/node_modules,

I can not figure out what the main issue is. I think the time those node modules gets copied to the output folder, rollup cant find devexpress which is requested by the qs-ui library?
But I actually dont know how to solve this issue.

devextreme-vue and devextreme is in both package.json under dependencies, so its there...

----

![image](https://github.com/unjs/nitro/assets/53934248/3bba58ca-c5d3-49b6-abd3-68f4263bfaea)

> https://github.com/rollup/rollup/discussions/5361

**Comment by skoenfaelt** [comment, marked as answer]
Ahhh i messed up...
In the nuxt config you can adress vite and nitro config. both have rollup options.

I thought the solution would be somewhere with that externals in the rollup options. In the qs-ui library i got this
```
external: ['vue', 'vuetify/lib', 'axios', 'devextreme', 'devextreme-vue/data-grid'],
```

and for the nuxt app its more or less equal but its meant for the nitro server
https://nitro.unjs.io/config#rollup

So to fix that issue, i have to name all external libs which make issues (i cant explain the "magic" behind it...)

Nuxt Config of the nuxt app
```
...
nitro: {
    rollupConfig: {
      external: ['devextreme', 'devextreme-vue/data-grid', 'devextreme/core/config']
    }
  },
...
```

with that build works.

**Reply by skoenfaelt** [reply]
In fact im the biggest piece of garbage in human existence :D
 
I just came to the issue, the built output of devextreme still uses cjs, which is wrong. But I just saw devextreme has now esm support with v23.2, so qs-ui dont need treat devextreme es external, and could built inside the esm code, which makes any external clause useless.

**Reply by pi0** [reply]
I haven't had a chance to read this thread but you are amazing that could solve it yourself!

(JS is the biggest garbage of human achievement, and we love it!)
